Add English translation for course 9 (#20)
moonbitlang/moonbit-course: course9/lecture_en.md

# Traits

<!--
```moonbit
enum Tree[T] {
  Empty
  Node(T, Tree[T], Tree[T])
}

struct Queue[T] {
  mut array: Array[T]
  mut start: Int
  mut end: Int
  mut length: Int
}
```
-->

## Methods

In Chapter 8, while studying mutable data structures, we discussed the need for providing a default value when constructing the array for implementing a generic circular queue.

```moonbit no-check
fn make[T]() -> Queue[T] {

> { array: Array::make(5, T::default()), start: 0, end: 0, length: 0 }

Reviewer: ```suggestion
  { array: Array::make(5, T::default()), start: 0, end: 0, length: 0 }
```

  match tree {
    Empty => Node(value, Empty, Empty) 
    Node(v, left, right) =>  
      if T::compare(value, v) == 0 { // We can call `compare` here.
        tree
      } else if T::compare(value, v) < 0 { // We can call `compare` here.
        Node(v, insert(left, value), right)
      } else {
        Node(v, left, insert(right, value))
      }
  }
}
```

## Implementation of Traits

Since the trait system of MoonBit is structured, to implement a trait, all we need to do is to define the corresponding methods using the syntax `fn <type>::<method>(...) -> ...`.

In the following example, we define the `default` method for the `BoxedInt` type.

```moonbit
struct BoxedInt { value : Int }

fn BoxedInt::default() -> BoxedInt {
  { value : Int::default() }
  // The default value can be defined by boxing the default value of `Int`.
}
```

By defining the `default` method, the `Default` trait is now implemented. Therefore, we can construct a queue of `BoxedInt` using the `make` function, which internally invokes the `default` method of `BoxedInt`.

```moonbit no-check
fn init {
  let array: Queue[BoxedInt] = make()
}
```

> 

Reviewer: There's another syntax:
```rust
// 为 `Trait` 的方法 `method` 提供默认实现
impl Trait with method(...) { ... }

// 给类型 `Type` 实现 `Trait` 的方法 `method`
impl Trait for Type with method(...) { ... }

// 带类型参数的情况
impl[X] Trait for Array[X] with method(...) { ... }
```
reference: https://www.moonbitlang.cn/weekly-updates/2024/04/30/weekly-04-30

## Method Chaining

In addition to `<type>::<method>(<expr>, ...)`, we can as well call the method using `<expr>.<method>(...)`, given `<expr>` is of type `<type>`. When defining such methods, `<type>::` can also be omitted when the first parameter is named `self`.

In this way, when we need to call such functions sequentially, we can avoid multiple nesting of method calls, thereby enhancing the readability of the code.

```moonbit
fn BoxedInt::plus_one(b: BoxedInt) -> BoxedInt {
  { value : b.value + 1 }
}

> fn plus_two(self: BoxedInt) -> BoxedInt {
>   // `<type>::` can be omitted when the first parameter is named `self`.

Reviewer: ```suggestion
// `<type>::` can be omitted when the first parameter is named `self`.
fn plus_two(self: BoxedInt) -> BoxedInt {  
```
The `<type>::` is omitted in the type signature, so it may be better if the order is adjusted

  { value : self.value + 2}
}

fn init {
  let _five = { value: 1 }.plus_one().plus_one().plus_two()
  // This avoids multiple nesting of method calls.
  let _five = plus_two(plus_one(plus_one({value: 1})))
}
```

## Automatic Derivation of Builtin Traits

In MoonBit, it is possible to automatically derive certain basic traits by adding `derive(<traits>)` after the type declaration. If the type is composed of other types, those member types must have already implemented the same traits.

> In the given example, we derive the `Default`, `Eq`, `Compare`, and `Debug` traits for `BoxedInt`. These traits can be successfully derived because `Int` already implements them. However, if the required traits are not implemented by the member types, the compiler will generate an error message.

Reviewer: TODO: `Debug` will be merged with `Show` some day, so we'll need to update this in the future

The map can be implemented using a list of pairs.

```moonbit
type Map[Key, Value] List[(Key, Value)]
```

The first two basic methods, `make` and `put`, can be easily implemented as follows:
- Create a map by creating an empty list.
  ```moonbit
  fn make[Key, Value]() -> Map[Key, Value] { 
    Map(Nil)
  }
  ```
- Add/update a key-value pair by inserting the pair to the beginning of the list.
  ```moonbit
  fn put[Key, Value](map: Map[Key, Value], key: Key, value: Value) -> Map[Key, Value] { 
    let Map(original_map) = map
    Map( Cons( (key, value), original_map ) )
  }
  ```

The third method, `get`, is also easy to describe in prose:
- Search the list from the beginning until the first matching key is found.

In such an implementation of the `get` function, we need to compare the key we are searching for with the keys stored in the map to determine if they are equal. As a result, the `Key` type must implement the `Eq` trait. That is, we need to modify the previous declaration of `get` so that `Key` is bound by `Eq`.

The complete implementation of `get` is as follows:

```moonbit
fn get[Key: Eq, Value](map : Map[Key, Value], key : Key) -> Option[Value] {

>   fn aux(list : List[(Key, Value)]) -> Option[Value] {
>     match list {
>       Nil => None
>       Cons((k, v), tl) => if k == key {
>         // `Key` is bound by `Eq`, so we can call `==` directly.
>         Some(v)
>       } else {
>         aux(tl)
>       }
>     }
>   }
> 
>   aux(map.0) // Use `.0` to get the value.
> }

Reviewer: ```suggestion
  loop map.0 {
    Nil => None
    Cons((k, v), tl) => if k == key {
      // `Key` is bound by `Eq`, so we can call `==` directly.
      Some(v)
    } else {
      continue tl
    }
  }
}
```

```

> In the above implementation, we define an auxiliary function called `aux` to traverse the list and use the `==` operator to determine whether the current key matches the one we are looking for. It can be shown that when there are multiple identical keys, we will always retrieve the latest value associated with that key.

Reviewer: ```suggestion
In the above implementation, we loop through the list and use the `==` operator to determine whether the current key matches the one we are looking for. It can be shown that when there are multiple identical keys, we will always retrieve the latest value associated with that key.
```

  let _ = { value: 10 } + { value: 100 } // { value: 110 }
}
```

In the previous example, we used the `get` and `set` functions to access and update data in the map. However, it is also possible to use custom operators to achieve the same functionality. By defining the `op_get` and `op_set` methods for the `Map` type, we can use the `map[k]` syntax to retrieve the value corresponding to `k`, and use `map[k] = v` to update the value associated with `k` to `v`.

```moonbit
fn Map::op_get[Key: Eq, Value](map: Map[Key, Value], key: Key) -> Option[Value] {
  get(map, key)
}
fn Map::op_set[Key: Eq, Value](map: Map[Key, Value], key: Key, value: Value) -> Map[Key, Value] {
  put(map, key, value)
}

fn init {
  let empty: Map[Int, Int] = make()
  let one = { empty[1] = 1 } // let one = Map::op_set(empty, 1, 1)
  let _ = one[1] // let _ = Map::op_get(one, 1)
}
```

In the previous example, the assignment statement is essentially an expression, and the return value is determined by the `op_get` method. We can use this method to retrieve the updated value. For mutable data structures, we also have the option to modify them in place without returning any value.

> 

Reviewer: Maybe we can mention the pattern matching support for `Map` here?
